Added newsletter from address verification

kevinansfield opened the pull request:
refs https://github.com/TryGhost/Team/issues/1498
refs https://github.com/TryGhost/Team/issues/584

With `membersFromAddress` being removed in favor of separate from addresses for each newsletter the verification flow needs to be moved too.

This also addresses problems with the existing verification flow where the emailed link was pointing directly at an API route resulting in poor experience when the process is unsuccessful. With this new flow there's no specific endpoint for triggering an email change, instead a verification will be triggered as needed when creating/updating a newsletter with `meta` data inserted in the response for clients to inform users as needed. The email also links to an Admin endpoint rather than an API endpoint providing a smoother confirmation flow.

- updated `newsletter` service so the index exports an instance
  - follows the more common pattern for services
  - keeps dependency injection in one place so it's not necessary to re-specify the injected dependencies everywhere the service is consumed
- added `add` and `edit` methods to `NewsletterService` class
  - accepts same arguments as `NewsletterModel.add/edit`
  - if the supplied `sender_email` attribute requires verification the attribute is stripped before saving, a verification email is triggered, and a `meta: {sent_email_verification: ['sender_email']}` property is added to the response
  - email verification contains a link to Admin, eg https://yoursite.com/ghost/#/settings/members-email/?verifyEmail=token with the token containing the id of the newsletter that's being updated, the new email value, and the changed property name
- updated `newsletters` API so it uses the `NewsletterService.add/edit` methods that contain the verification email trigger business logic
- added `PUT /newsletters/verifications/` endpoint
  - accepts `{token: 'xyz'}` in the body, this request is triggered by Admin when the confirmation link is loaded
  - extracts data from the single-use token, then uses the `id`, `property`, and `value` data to edit the newsletter completing the update of the email address

TODO:
- [x] add/edit tests
- [x] verification tests

kevinansfield committed:
Added newsletter from address verification

refs https://github.com/TryGhost/Team/issues/1498
refs https://github.com/TryGhost/Team/issues/584

With `membersFromAddress` being removed in favor of separate from addresses for each newsletter the verification flow needs to be moved too.

This also addresses problems with the existing verification flow where the emailed link was pointing directly at an API route resulting in poor experience when the process is unsuccessful. With this new flow there's no specific endpoint for triggering an email change, instead a verification will be triggered as needed when creating/updating a newsletter with `meta` data inserted in the response for clients to inform users as needed. The email also links to an Admin endpoint rather than an API endpoint providing a smoother confirmation flow.

- updated `newsletter` service so the index exports an instance
  - follows the more common pattern for services
  - keeps dependency injection in one place so it's not necessary to re-specify the injected dependencies everywhere the service is consumed
- added `add` and `edit` methods to `NewsletterService` class
  - accepts same arguments as `NewsletterModel.add/edit`
  - if the supplied `sender_email` attribute requires verification the attribute is stripped before saving, a verification email is triggered, and a `meta: {sent_email_verification: ['sender_email']}` property is added to the response
  - email verification contains a link to Admin, eg https://yoursite.com/ghost/#/settings/members-email/?verifyEmail=token with the token containing the id of the newsletter that's being updated, the new email, and the changed property
- updated `newsletters` API so it uses the `NewsletterService.add/edit` methods that contain the verification email trigger business logic
- added `PUT /newsletters/verify-email/` endpoint
  - accepts `{token: 'xyz'}` in the body, this request is triggered by Admin when the confirmation link is loaded
  - extracts data from the single-use token, then uses the `id`, `email`, and `property` data to edit the newsletter completing the update of the email address

ErisDS commented:
I have some comments on naming 😬  ref: [API design docs](https://www.notion.so/ghost/API-Core-c7f1926386304eedb3d8fd4304bfecf8)

- endpoint names should be snake_case not kebab
- endpoint names should be nouns wherever possible :) 


Suggestions: 

`PUT /newsletters/emails/`
`PUT /newsletters/verifications/`
`PUT /newsletters/from_address/`


Any of these or similar would be fine, depends on how you think about the resource that you're operating on. 

In terms of existing patterns in the API that relate to this

`router.put('/authentication/passwordreset'` is 👍 
`router.post('/authentication/invitation'`  is 👍 (although plural would prob be even better)
`router.post('/authentication/reset_all_passwords'` is 👎 

Also (I've spent way to long looking at the request and response bodies of all our endpoints recently haha) if we have no strong use-case for a response body from the endpoint, I am heavily in favour of setting `statusCode: 204` for any custom endpoint that does a thing and reports OK or errors. I can't quite tell from the PR if there's a real need for returning the edit response?

kevinansfield committed:
dropped unused `config` injection

kevinansfield committed:
added tests

kevinansfield commented:
> I have some comments on naming 😬 ref: [API design docs](https://www.notion.so/ghost/API-Core-c7f1926386304eedb3d8fd4304bfecf8)

👍 I'll update to `PUT /newsletters/verifications/`. The token contains the property name and the new value so it's a fairly generic property/value verification process.

> I can't quite tell from the PR if there's a real need for returning the edit response?

The response is by the Admin as it lets it update it's internal store with the newly updated data. Clients can't read the data from the token, all they knows is that it needs to send it to the API to get the updated newsletter in response. An alternative would be returning something minimal with the `id` of the newsletter so the client can re-fetch it to get the updated version 🤔

kevinansfield committed:
switched to `urlUtils.urlFor('admin', true)`

kevinansfield committed:
fixed linting

kevinansfield committed:
fixed unit test

kevinansfield committed:
fixed duplicated `/ghost/` in email update magic link

kevinansfield committed:
renamed verify route to `PUT /newsletters/verifications/`

kevinansfield committed:
added email verification confirmation test

kevinansfield committed:
fixed linting

ErisDS commented:
Ok, that all makes sense to me :)

ErisDS committed:
Removed unused settingsCache from newsletterService

- This reference isn't used anywhere so I've removed it

ErisDS committed:
Added unit tests for newsletterservice public methods

- We're really not that far away from being able to use private class methods
- Therefore, we should follow the pattern of only writing unit tests for public class methods
- These tests give us near 100% coverage of private class methods
- Need to finish testing verifyPropertyUpdate but that requires a small refactor

ErisDS committed:
Added full verifyPropertyUpdate tests + refactor

- Refactored the way the singleUseTokenProvider is setup so that we can override in tests
- Moved verifyPropertyUpdate to the public section of the newsletter methods
- Wired up a test to show that the function passes data from the token in the expected format
